typing/typemod.ml

```diff
@@ -2485,16 +2485,17 @@ and type_structure ?(toplevel = false) funct_body anchor env sstr =
           (fun rs info -> Sig_type(info.typ_id, info.typ_type, rs, Exported))
           decls []
         in
-        let shape_map = List.fold_left
-          (fun shape_map -> function
-            | Sig_type (id, vd, _, _) ->
+        let shape_map = List.fold_left2
+          (fun shape_map type_decl  -> function
+            | Sig_type (id, td, _, _) ->
               if not (Btype.is_row_name (Ident.name id)) then begin
-                Env.register_uid vd.type_uid vd.type_loc;
-                Shape.Map.add_type shape_map id vd.type_uid
+                Env.register_uid td.type_uid type_decl.typ_name.loc;
+                Shape.Map.add_type shape_map id td.type_uid
               end else shape_map
             | _ -> assert false
           )
           shape_map
+          decls
           items
```

Comment:
You don't need to keep iterating on the `items` if you have `decls` available. `id` above is just `type_decl.type_id` and `vd/td` is `type_decl.typ_type`.